turludock/generate_dockerfile.py

```diff
@@ -0,0 +1,245 @@
+
+    # Base image
+    dockerfile += generate_from(yaml_config)
+
+    # Meta data
+    dockerfile += generate_header_info(_generate_description(yaml_config), yaml_config["ros_version"])
+
+    # Common configuration for all images
+    dockerfile += generate_common_env_config()
+    dockerfile += generate_install_common_packages()
+    dockerfile += generate_locale()
+    dockerfile += generate_cmake(_get_package_version(yaml_config, "cmake"))
+    dockerfile += generate_terminator()
+    dockerfile += generate_ohmyzsh()
+
+    # GPU driver
+    ubuntu_version = get_ubuntu_version(yaml_config["ros_version"])
+    if yaml_config["gpu_driver"] == "mesa":
+        if is_version_lower(ubuntu_version["semantic"], "20.04"):
+            logger.warning(
+                "Using mesa driver. However latest mesa drivers cannot be installed for "
+                + f"Ubuntu {ubuntu_version['semantic']}. This means if you use laster Intel/AMD GPU "
+                + "you most likely not going to get HW acceleration and programs like RViZ will use "
+                + "software rendering. For more info check ppa:kisak/kisak-mesa"
+            )
+            # use mesa from default ubuntu repos
+            dockerfile += generate_mesa(False)
+        elif is_version_greater(ubuntu_version["semantic"], "22.04"):
+            # use mesa from default ubuntu repos - those should be the latest
+            dockerfile += generate_mesa(False)
+        else:
+            # use latest mesa provided by custom ppa
+            dockerfile += generate_mesa(True)
+    else:
+        # no need to install something for NVIDIA
+        pass
+
+    # Add CUDA/cuDNN
+    ubuntu_version = get_ubuntu_version(yaml_config["ros_version"])
+    if "cuda_version" in yaml_config:
+        dockerfile += generate_cuda_base(yaml_config["cuda_version"], ubuntu_version["flat"])
+        dockerfile += generate_cuda_devel(yaml_config["cuda_version"], ubuntu_version["flat"])
+        dockerfile += generate_cuda_runtime(yaml_config["cuda_version"], ubuntu_version["flat"])
+        if "cudnn_version" in yaml_config:
+            dockerfile += generate_cudnn_devel(
+                yaml_config["cuda_version"], yaml_config["cudnn_version"], ubuntu_version["flat"]
+            )
+            # TODO(ATA): Devel already contains what runtime has. So it redundant.
+            # But is it the case for cuDNN implementation?
+            dockerfile += generate_cudnn_runtime(
+                yaml_config["cuda_version"], yaml_config["cudnn_version"], ubuntu_version["flat"]
+            )
+
+    # Add ROS
+    dockerfile += generate_ros(yaml_config["ros_version"])
+
+    # Add the extra-packages
+    extra_packages_label_list = list()
+    if "extra_packages" in yaml_config:
+        for item in yaml_config["extra_packages"]:
+            if isinstance(item, dict):
+                package_name = next(iter(item), None)
+            elif isinstance(item, str):
+                package_name = item
```

Comment:
No else statement. Are you sure it is not needed?

Reply:
It should never happen. Before this is called we anyway do a sanity check for the yaml configuration. So at this point, it can only be a `str` or a `dict`.
Should I put an `else` with a `raise`?